src/Components/ADT3DSceneBuilder/Internal/Elements/ElementForm.tsx

```diff
@@ -143,259 +121,334 @@ const SceneElementForm: React.FC<IADT3DSceneBuilderElementFormProps> = ({
                     );
+                    logDebugConsole(
+                        'debug',
+                        `Adding element ${formElementId} to behavior ${behavior.displayName} (${behavior.id}).`
+                    );
 
                     // add the behavior to the current scene if it is not there
-                    updatedConfig = ViewerConfigUtility.addBehaviorToScene(
+                    ViewerConfigUtility.addBehaviorToScene(
                         updatedConfig,
                         sceneId,
+                        behavior,
+                        true // in-place update
+                    );
+                } else {
+                    ViewerConfigUtility.removeElementFromBehavior(
+                        formElementId,
                         behavior
                     );
+                    logDebugConsole(
+                        'debug',
+                        `Removing element ${formElementId} from behavior ${behavior.displayName} (${behavior.id}).`
+                    );
+
+                    // TODO: should we remove the behavior if it no long maps to any elements?
                 }
-            }
+            });
             // END of behaviors update which this element exists in
-
+            logDebugConsole(
+                'info',
+                '[END] Saving element form. {finalConfig, element, linkedBehaviors}',
+                updatedConfig,
+                elementFormState.elementToEdit,
+                elementFormState.linkedBehaviorIds
+            );
             return adapter.putScenesConfig(updatedConfig);
         },
         refetchDependencies: [adapter],
         isAdapterCalledOnMount: false
     });
 
-    const handleSaveElement = useCallback(async () => {
-        await saveElementAdapterData.callAdapter();
+    // callbacks
+    const handleSaveElement = useCallback(() => {
+        saveElementAdapterData.callAdapter();
     }, [saveElementAdapterData]);
 
+    const notifySceneContextDirtyState = useCallback(
+        (isDirty: boolean) => {
+            dispatch({
+                type: SET_ADT_SCENE_BUILDER_FORM_DIRTY_MAP_ENTRY,
+                payload: {
+                    formType: 'element',
+                    value: isDirty
+                }
+            });
+        },
+        [dispatch]
+    );
+
+    const onCancelClick = useCallback(() => {
+        if (elementFormState.isDirty) {
+            setUnsavedBehaviorChangesDialogOpen(true);
+            setUnsavedChangesDialogDiscardAction(onElementBackClick);
+        } else {
+            onElementBackClick();
+        }
+    }, [
+        elementFormState.isDirty,
+        onElementBackClick,
+        setUnsavedBehaviorChangesDialogOpen,
+        setUnsavedChangesDialogDiscardAction
+    ]);
+
     const handleCreateBehavior = useCallback(
         async (searchText: string) => {
             // Save element
-            await handleSaveElement();
+            handleSaveElement();
```

Comment:
Why is this no longer async?

Reply:
okay so this is actually a little scary, like if we create an element, we want to make sure that writing the config file succeeds BEFORE we go to the elements list, because the content of the elements list may be populated from reading the config file from storage

Reply:
reverted